Assets/Scripts/GameStateMachine/State.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+public abstract class State
+{
+    private State _nextState;
+
+    protected GameStateMachineBehaviour OwningGameStateMachine { get; private set; }
+
+    public Action OnStateEnter;
+    public Action OnStateExit;
+
+    protected State() => InitState();
+
+    protected void InitState()
+    {
+        this.Subscribe();
+        OwningGameStateMachine = GameStateMachine.OwningGameStateMachine;
+    }
+    
+    protected void SetNextState<TState>() where TState: State, new() 
+        => _nextState = GameStateMachine.GetState<TState>();
```

Comment:
I'm just wondering how the syntax here works. If you can explain that would be great.

Reply:
The method is generic, as indicated by `<TState>`, which means it can work with any type that is or inherits from `State`. The `where TState: State, new()` LINQ expression constraint ensures that `TState` is or derives from `State` and has a parameterless constructor. This is necessary because the method needs to create an instance of `TState`.